stacks-signer/src/chainstate.rs

```diff
@@ -147,12 +150,40 @@ impl<'a> ProposedBy<'a> {
 impl SortitionsView {
     /// Apply checks from the SortitionsView on the block proposal.
     pub fn check_proposal(
-        &self,
+        &mut self,
         client: &StacksClient,
         signer_db: &SignerDb,
         block: &NakamotoBlock,
         block_pk: &StacksPublicKey,
     ) -> Result<bool, SignerChainstateError> {
+        // If this is the first block in the tenure, check if it was proposed after the timeout
+        if signer_db
+            .get_last_signed_block_in_tenure(&block.header.consensus_hash)?
+            .is_none()
+        {
+            if let Some(received_ts) =
+                signer_db.get_burn_block_receive_time(&self.cur_sortition.burn_block_hash)?
+            {
+                let received_time = UNIX_EPOCH + Duration::from_secs(received_ts);
+                let elapsed = std::time::SystemTime::now()
+                    .duration_since(received_time)
+                    .unwrap_or_else(|_| {
+                        panic!("Failed to calculate time since burn block received")
+                    });
+                if elapsed >= self.config.block_proposal_timeout {
+                    warn!(
+                        "Miner proposed first block after block proposal timeout.";
+                        "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                        "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                        "current_sortition_consensus_hash" => ?self.cur_sortition.consensus_hash,
+                        "last_sortition_consensus_hash" => ?self.last_sortition.as_ref().map(|x| x.consensus_hash),
+                        "burn_block_received_time" => ?received_time,
+                    );
+                    self.cur_sortition.miner_status =
+                        SortitionMinerStatus::InvalidatedBeforeFirstBlock;
```

Comment:
I would add a `return false` here as well.

Reply:
But isn't it possible that it is still valid? what if the proposal is delayed but builds off the prior sortition?

Reply:
Late arrival is enough to be considered "invalid" -- this discourages miners from delaying their proposals. If proposals arrive late, it increases the pressure on the network to get blocks confirmed before the next bitcoin block (and other miners who are competing in the next block would need to be able to get their block commits in time).

Reply:
Then what is the point of setting this sortition state and then later checking if the block is build off the last sortition?